### PR TITLE
Fix memory leak

### LINKED ISSUE #35: Mem leak in level1c_runner
level1c_runner has a mem leak.

After running for 24 hours without restart top gives like:

```
    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
 345356 ubuntu    20   0   11.7g   3.3g  11428 S   0.0  10.5  42:11.76 /software/miniconda/envs/ppsv2021/bin/python /software/miniconda/envs/ppsv2021/bin/level1c_runner.py -vv -s viirs-l1c -c /software/+
```

This is on a 32GB system

### PR DESCRIPTION
**Pull request by ninahakansson**
Replaces the ThreadPool with a simple non parallel Process. We saw that the memory usage was increasing over time. Also the ThreadPool were never sent more than one message at a time.

 - [x] Closes #35 
 - [x] Tests added 
 - [X] Tests passed: Passes ``pytest nwcsafpps_runner`` 
 - [X] Passes ``flake8 nwcsafpps_runner`` 
 - [ ] Fully documented

### CODE REVIEW

**Comment by mraspaud** [comment]
Looks like the tests are failing, is this expected?

**Comment by adybbroe** [comment]
Tests are most likely failing because Python 3.7 is getting dropped from conda-forge. See #52

File: bin/level1c_runner.py [hunk 1]
```diff
@@ -53,28 +52,26 @@ def signal_handler(sig, frame):
...
-            LOG.info("L1C processing has completed.")
+            if l1c_proc.l1cfile is not None:
+                pub_msg = prepare_l1c_message(l1c_proc.l1cfile,
+                                              l1c_proc.message_data,
+                                              orbit=l1c_proc.orbit_number)
+                publish_l1c(publisher, pub_msg,
+                            publish_topic=l1c_proc.publish_topic)
+                LOG.info("L1C processing has completed.")
+            else:
+                LOG.debug("L1C processing has failed.")
```
Comment (adybbroe) [inline bin/level1c_runner.py:74]: Shouldn't this rather be a warning?

**Review by adybbroe — COMMENTED** [review]
Thanks! I have only looked rather quickly, and is fine so far with your changes. Only two comments/questions. And maybe you can check the codecov complaints, eventhough it is only 0.1% decrease ...

**Comment by ninahakansson** [comment]
Jag håller inte med coveralls om att den raden inte är testad.

File: nwcsafpps_runner/l1c_processing.py [hunk 2]
```diff
@@ -142,10 +144,24 @@ def run(self, msg):
...
+        process1.start()
+        process1.join(self.time_limit_seconds)  # Normally takes 3-8s for VIIRS
+        self.l1cfile = result_dict.get("l1cfile", None)
+        try:
+            process1.close()
+        except ValueError:
+            # if process1.is_alive():
+            LOG.warning(
+                "Processing level1c file not terminated after 60s.")
+            process1.terminate()
```
Comment (TAlonglong) [inline nwcsafpps_runner/l1c_processing.py:164]: So, I'm not sure I follow this. But I think the idea is to start each new processing in a separate process. Without any limitation than processing time. I think that quickly can get you into problems if your processing is using more memory than you have available.

File: nwcsafpps_runner/tests/test_level1c_runner.py [hunk 3]
```diff
@@ -328,14 +323,25 @@ def test_orbit_number_from_msg_viirs(self, cpu_count, config):
...
-                l1c_proc.run(input_msg)
+        with tempfile.NamedTemporaryFile() as myconfig_file:
+            l1c_proc = L1cProcessor(myconfig_file.name, 'viirs-l1c')
+            l1c_proc.run(input_msg)
 
         expected_orbit_number = TEST_VIIRS_MSG_DATA.get('orbit_number')
         self.assertEqual(l1c_proc.orbit_number, expected_orbit_number)
 
+    @patch('nwcsafpps_runner.config.load_config_from_file')
+    @patch.dict('nwcsafpps_runner.l1c_processing.LVL1C_PROCESSOR_MAPPING', {'viirs-l1c': my_fake_l1proc_function})
```
Comment (TAlonglong) [inline nwcsafpps_runner/tests/test_level1c_runner.py:334]: Are you really sure you patch the correct place?

File: nwcsafpps_runner/l1c_processing.py (hunk 2 above)
Comment (ninahakansson) [reply nwcsafpps_runner/l1c_processing.py:164]: It starts a new process but also waits for it to finish (with join). So there is never more than one active process. The limit is just to clean up if a process hangs for any reason. In the end I made the time limit configurable with 60s as default. (I will fix the message.)

File: nwcsafpps_runner/tests/test_level1c_runner.py (hunk 3 above)
Comment (ninahakansson) [reply nwcsafpps_runner/tests/test_level1c_runner.py:334]: The right place as in "is the function actually run by the test"? Yes, I am certain of that. If I change the time_limit_seconds: 1 to a higher value the test takes much longer as it is actually running the function sleep(65) and is interrupted only after time_limit_seconds. But there might be a better place to patch it though. I tried several worse places to patch before I came up with this one. For all other places I tried the my_fake_l1proc_function was never interrupted.

**Review by TAlonglong — APPROVED** [review]
As far as I can see this LGTM

**Comment by TAlonglong** [comment]
> Jag håller inte med coveralls om att den raden inte är testad.

OK, I think I see what happens. The fake sleep starts, but the function is killed before the time.sleep finishes and then the coveralls to cover that line.

But I don't see an easy solution

**Comment by mraspaud** [comment]
> > Jag håller inte med coveralls om att den raden inte är testad.
> 
> OK, I think I see what happens. The fake sleep starts, but the function is killed before the time.sleep finishes and then the coveralls to cover that line.
> 
> But I don't see an easy solution

Yes, I agree. An ugly solution would be to replace `time.sleep(65)` with
```python
for i in range(650):
    time.sleep(0.1)
```
but I think I'd rather have that line counted as not tested :)

**Pull request merged**